packages/scene-graph-pixi/runtime/src/importer/component/RichText.ts

```diff
@@ -0,0 +1,212 @@
+            width: width ? parseInt(width, 10) : 0
+          }
+        } as RichTextOutlineStyle;
+        break;
+      }
+      default: break;
+    }
+
+    return null;
+  }
+
+  /**
+   *
+   */
+  private static collectChildNodes(
+    nodes: NodeListOf<ChildNode>,
+    fragments: RichTextFragment[] = [],
+    styles: RichTextStyle[] = []
+  ): RichTextFragment[] {
+    for (let i = 0; i < nodes.length; i++) {
+      const node = nodes[i] as HTMLElement;
+      const style = RichText.parseNodeStyle(node);
+
+      if (style) {
+        styles.push(style);
+      }
+
+      if (node.childNodes.length > 0) {
+        RichText.collectChildNodes(node.childNodes, fragments, styles);
+      }
+
+      if (node.nodeValue) {
+        fragments.push({ text: node.nodeValue, styles: styles.slice(0) });
+      }
+
+      if (style) {
+        styles.pop();
+      }
+    }
+
+    return fragments;
+  }
+
+  private static parseBBCode(input: string): RichTextFragment[] {
+    const richTextHtml = `<${RichText.PARSER_TAG_NAME}>${input}</${RichText.PARSER_TAG_NAME}>`;
+    const replacedRichTextString = richTextHtml.replace(/<([a-zA-Z]+)=([^\s]+)/g, '<$1 $1=$2');
+
+    const tempHtml = document.createElement('html');
+    tempHtml.innerHTML = replacedRichTextString;
+    const tags = tempHtml.getElementsByTagName('*');
+
+    let offset = 0;
+    let richTextTag = tags[offset];
+    while (richTextTag && richTextTag.tagName.toUpperCase() !== RichText.PARSER_TAG_NAME) {
+      richTextTag = tags[++offset];
+    }
+
+    return RichText.collectChildNodes(richTextTag.childNodes);
+  }
+
+  private static pixiTextStyleOptionsByFragment(fragment: RichTextFragment): PIXI.TextStyleOptions {
+    const params: PIXI.TextStyleOptions = {};
+    for (let j = 0; j < fragment.styles.length; j++) {
+      const style = fragment.styles[j];
+      switch (style.name) {
+        case 'B':     params.fontWeight = 'bold'; break;
+        case 'I':     params.fontStyle = 'italic'; break;
+        case 'SIZE':  params.fontSize = parseInt(style.params.value, 10); break;
+        case 'COLOR': params.fill = [style.params.value]; break;
+        case 'OUTLINE': {
+          params.stroke = style.params.color;
+          params.strokeThickness = parseInt(style.params.width, 10);
+          break;
+        }
+      }
+    }
+
+    return params;
+  }
+
+  private static createRichTextContainer(
+    fragments: RichTextFragment[],
+    defaultParams: PIXI.TextStyleOptions = {}
+  ): PIXI.Container {
+    const container = new PIXI.Container();
+
+    let x = 0;
+    let y = 0;
+    let lastLineHeight = 0;
+    for (let i = 0; i < fragments.length; i++) {
+      const fragment = fragments[i];
+      const params = RichText.pixiTextStyleOptionsByFragment(fragment);
+      const defaultParamsClone = Object.assign({}, defaultParams);
+      const style = new PIXI.TextStyle(Object.assign(defaultParamsClone, params));
+      const lines = fragment.text.split('\n');
```

Comment:
I think that PIXI will line breaks by '\n', but why is this process necessary?

Reply:
@zprodev 
Let's see this case;

<img width="105" alt="figure" src="https://user-images.githubusercontent.com/2690685/59832669-6704ac80-937f-11e9-88fb-c1bd74b09c47.png">

The first line has characters with different size.
If new line was began with the position based on the size of tail character, the head of the second line will intersects with the character on first line.
So this method does calibration of the line height with the largest character in the line.

Reply:
I see, LGTM.